tools/dds/dds-sniffer/dds-sniffer.cpp

```diff
@@ -0,0 +1,302 @@
+
+using namespace TCLAP;
+
+int main( int argc, char ** argv ) try
+{
+    eprosima::fastdds::dds::DomainId_t domain = 0;
+    uint32_t seconds = 0;
+
+    CmdLine cmd( "librealsense rs-dds-sniffer tool", ' ' );
+    SwitchArg snapshot_arg( "s", "snapshot", "run momentarily taking a snapshot of the domain" );
+    ValueArg< eprosima::fastdds::dds::DomainId_t > domain_arg( "d",
+                                                               "domain",
+                                                               "select domain ID to listen on",
+                                                               false,
+                                                               0,
+                                                               "0-232" );
+    cmd.add( snapshot_arg );
+    cmd.add( domain_arg );
+    cmd.parse( argc, argv );
+
+    if( snapshot_arg.isSet() )
+    {
+        seconds = 3;
+    }
+
+    if( domain_arg.isSet() )
+    {
+        domain = domain_arg.getValue();
+        if( domain > 232 )
+        {
+            std::cerr << "Invalid domain value, enter a value in the range [0, 232]" << std::endl;
+            return EXIT_FAILURE;
+        }
+    }
+
+    Sniffer snif;
+    if( snif.init( domain ) )
+    {
+        snif.run( seconds );
+    }
+    else
+    {
+        std::cerr << "Initialization failure" << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    return EXIT_SUCCESS;
+}
+catch( const TCLAP::ExitException & )
+{
+    std::cerr << "Undefined exception while parsing command line arguments" << std::endl;
+    return EXIT_FAILURE;
+}
+catch( const TCLAP::ArgException & e )
+{
+    std::cerr << e.what() << std::endl;
+    return EXIT_FAILURE;
+}
+catch( const std::exception & e )
+{
+    std::cerr << e.what() << std::endl;
+    return EXIT_FAILURE;
+}
+
+Sniffer::Sniffer()
+    : _participant( nullptr )
+    , _subscriber( nullptr )
+    , _topics()
+    , _readers()
+    , _datas()
+    , _subscriberAttributes()
+    , _readerQoS()
+{
+}
+
+Sniffer::~Sniffer()
+{
+    _running = false;
+
+    for( const auto & it : _topics )
+    {
+        if( _subscriber != nullptr )
+        {
+            _subscriber->delete_datareader( it.first );
+        }
+        _participant->delete_topic( it.second );
+    }
+    if( _subscriber != nullptr )
+    {
+        if( _participant != nullptr )
+        {
+            _participant->delete_subscriber( _subscriber );
+        }
+    }
+    if( _participant != nullptr )
+    {
+        eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->delete_participant( _participant );
+    }
+
+    _datas.clear();
+    _topics.clear();
+    _readers.clear();
+}
+
+bool Sniffer::init( eprosima::fastdds::dds::DomainId_t domain )
+{
+    eprosima::fastdds::dds::DomainParticipantQos participantQoS;
+    participantQoS.name( "dds-sniffer" );
+    _participant = eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->create_participant( domain,
```

Comment:
So I still don't understand the error handling here:
Fastdds will return `nullptr` if something fails? Where's the error code? How do we know what failed? Is there anything?

Reply:
In this case fastDDS only returns nullptr on failure